app/models/company_update.rb

```diff
@@ -64,17 +65,78 @@ def validate_partners_values
-    return if company_values.is_a?(Array) && !company_values.detect { |value| !value.is_a?(Hash) }
-
-    errors.add(:company_values, :invalid)
-  end
-
   def validate_values_presence
     return if company_values || partners_values
 
     errors.add(:company_values, 'Company_values e Partners_values não podem ser ambos nulos')
     errors.add(:partners_values, 'Company_values e Partners_values não podem ser ambos nulos')
   end
+
+  def self.csv_columns(max_partners)
+    attributes = ['CNPJ', 'Nome', 'Razão social da empresa',
+                  'Ano de fundação', 'CNAE', 'Emails', 'Endereço', 'Bairro',
+                  'Cidade sede', 'Estado', 'CEP', 'Breve descrição', 'Site', 'Tecnologias',
+                  'Produtos e serviços', 'Objetivos de Desenvolvimento Sustentável',
+                  'Redes sociais', 'Número de funcionários contratados como CLT',
+                  'Número de colaboradores contratados como Pessoa Jurídica (PJ)',
+                  'Número de estagiários/bolsistas contratados',
+                  'A empresa está ou esteve em alguma incubadora ou Parque tecnológico',
+                  'A empresa recebeu investimento?', 'Investimentos',
+                  'Valor do investimento próprio (R$)', 'Valor do investimento-anjo (R$)',
+                  'Valor do Venture Capital (R$)', 'Valor do Private Equity (R$)',
+                  'Valor do PIPE-FAPESP (R$)', 'Valor de outros investimentos (R$), Financiamento',
+                  'Deseja a marca DNAUSP?', 'Nome', 'Email']
+
+    attributes.concat(['Nome do sócio', 'Email', 'Vínculo', 'Unidade', 'NUSP'] * max_partners)
+  end
+
+  def self.get_value(value)
+    value.nil? ? '' : value
+  end
+
+  def self.fetch_attributes_data_from(attributes, from)
+    attributes.map do |attr|
+      value = if from.is_a?(Hash)
+                from[attr]
+              else
+                from.send(attr)
+              end
+      get_value(value)
+    end
+  end
+
+  def self.fetch_partners_data_from(max_partners, attributes, from)
+    result = []
+    partners = get_value(from)
+    (0..max_partners - 1).each do |i|
+      if i < partners.length
+        partner = partners[i]
+        result.concat(fetch_attributes_data_from(attributes, partner))
+      else
+        result.concat([''] * 5)
+      end
+    end
+    result
+  end
+
+  def self.to_csv
+    max_partners = ENV['CSV_MAX_PARTNERS'] || 5
+    attributes = csv_columns(max_partners)
+    CSV.generate(headers: true) do |csv|
+      csv << attributes
+
+      all.each do |company|
+        row = []
+
+        row.concat(fetch_attributes_data_from(%w[cnpj name], company))
+        company_attrs = attributes[2..29]
+        row.concat(fetch_attributes_data_from(company_attrs, company.send('company_values')))
+        row.concat(fetch_attributes_data_from(%w[wants_dna name email], company.send('dna_values')))
+        row.concat(fetch_partners_data_from(max_partners, %w[name email bond unity nusp],
```

Comment:
Confesso que não to gostando muito do uso do `fetch` aqui. Você não tá buscando de outro lugar, você tá transformando dados que você já tem, certo?
Além disso, `attributes_data` parece redundante.
O que as funções estão fazendo, em um nível de abstração mais alto? Isso vai ajudar a nomeá-las melhor.